fairscale/optim/oss.py

```diff
@@ -334,12 +321,97 @@ def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        sharded properties are not exposed. It contains two entries:
+
+        * state - a dict holding current optimization state. Its content
+            differs between optimizer classes.
+
+        * param_groups - a dict containing all parameter groups
+
+        .. warning:
+            If the state has not been consolidated, this returns a shard's worth, not the global state.
+
+        .. warning:
+            Returning the global state is limited to the replica which was responsible for the consolidation.
+            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
+        """
+
+        if len(self._all_states) == 0:
+            logging.warning("Optimizer state has not been consolidated. Returning the local state")
+            logging.warning("Please call `consolidate_state_dict()` beforehand if you meant to save the global state")
+            return self.optim.state_dict()
+
+        # Unify the shard states and the state that pytorch would expect, given the model.
+        # Indexation needs several redirections, since each shard only knows a limited scope of the model
+        # - get the pytorch compliant parameter indexing
+        state_dict = super().state_dict()
+
+        # - get an id map which links the parameter id to the index in the reference state
+        global_id_map = [{id(p): i for pg in self.param_groups for i, p in enumerate(pg["params"])}]
+
+        # - go through the per-shard states, which are all indexed locally
+        for rank, s in enumerate(self._all_states):
+            # -- match the local indexing and the global partition, update the corresponding saved state globally
+            for local_pg, global_pg, rank_map in zip(
+                s["param_groups"], self.partition_parameters()[rank], global_id_map
+            ):
+                # Go through the parameters indexed locally, pick up the global corresponding param
+                # NOTE: Contents of the state_dict changes in between torch1.5 and torch1.6+
+                local_index_to_param_id = {
+                    i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
+                }
+
+                for local_param_index in local_pg["params"]:
+                    global_index = rank_map[local_index_to_param_id[local_param_index]]
+
+                    # Update the state, if any
+                    if local_param_index in s["state"].keys():
+                        state_dict["state"][global_index] = s["state"][local_param_index]
+
+        return state_dict
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Restore the global parameter groups as well as the shard.
+
+        Arguments:
+            state_dict (dict): optimizer state. Should be an object returned
+                from a call to :meth:`state_dict`
+        """
+
+        super().load_state_dict(state_dict)
```

Comment:
This is happening on CPU, right? For large models, we might get CPU OOM - e.g. if there are 8 processes (for 8 GPUs). One possible improvement (maybe you are doing this already) is to load on one rank only and then broadcast to the other ranks. Here's an example of how we do this in Fairseq: https://github.com/fairinternal/fairseq-py/blob/master/fairseq/trainer.py#L328

Reply:
good question !
- this call should not change the localization as far as I know, so if the state dict that it gets is already on different gpus for instance then the params will stay there. 
- I'll have a second look at the optim state here, I think you're right in that this will copy too much of a state on each rank -unused-, I can prune the unused states for this rank first

Reply:
If we load a pytorch checkpoint that hasn't been sharded, then it's not guaranteed to fit on a single GPU. How about we load on CPU, shard, then move to GPU?
Awesome, pruning the unused states could save a lot of memory for large DDP rank

Reply:
> How about we load on CPU, shard, then move to GPU?
when and who does that, on which gpu ? I don't think that there's the needed information at this level, or at least I'm not sure how that will look like. Here people can use `load_state_dict` with a state on CPU, that's completely fine, but I would not make that call within the optimizer, or at least I think that would break the logic for many people

> Awesome, pruning the unused states could save a lot of memory for large DDP rank
done, except for torch1.5 of course, which is different.. incoming fix